File: src/main/java/com/rfb/demo/rxjavatest/algorithm/leetCode2/characterReplacement.java
package com.rfb.demo.rxjavatest.algorithm.leetCode2;

public class characterReplacement {

    public int characterReplacement(String s, int k) {

        int[] buck = new int[26];
        int left = 0;
        int max = 0;
        int n = s.length();
        for(int right = 0; right < n; right++){
            int index = s.charAt(right) - 'A';
            buck[index]++;

            if(max < buck[index]){
                max = buck[index];
            }

            if(right - left + 1 > max+k){
                buck[s.charAt(left)-'A']--;
                left++;
            }
        }

        return n-left;
    }

    public void test(){
        System.out.println(characterReplacement("BAAA", 0));
    }
}

File: src/main/java/com/rfb/demo/rxjavatest/algorithm/leetcode4/characterReplacement.kt
package com.rfb.demo.rxjavatest.algorithm.leetcode4

class characterReplacement {

    fun characterReplacement(s: String, k: Int): Int {

        var max = 0
        for(i in 0..25){

            val c = 'A' + i

            var rest = k
            var left = 0
            var right = 0
            while(right < s.length){
                if(s[right] == c){
                    right++
                }else{
                    if(rest > 0){
                        rest--
                        right++
                    }else{
                        if(s[left]  != c){
                            rest++
                        }
                        left++
                    }
                }

                if((right - left) > max){
                    max = right - left
                }
            }
        }
        return max
    }
    fun test(){
        println(characterReplacement("AABABBA", 1));
    }
}